File: src/NBdbm/Classes/Fachada_Net/command.vb
Option Explicit On 

Imports System
Imports System.Data

Namespace DotNetDataProviderTemplate
    Public Class TemplateCommand
        Implements IDbCommand

        Private m_connection As TemplateConnection
        Private m_txn As TemplateTransaction
        Private m_sCmdText As String
        Private m_updatedRowSource As UpdateRowSource = UpdateRowSource.None
        Private m_parameters As TemplateParameterCollection = New TemplateParameterCollection

        ' Implement the default constructor here.
        Public Sub New()
            MyBase.New()
        End Sub

        ' Implement other constructors here.
        Public Sub New(ByVal cmdText As String)
            MyBase.New()

            m_sCmdText = cmdText
        End Sub

        Public Sub New(ByVal cmdText As String, ByVal connection As TemplateConnection)
            MyBase.New()

            m_sCmdText = cmdText
            m_connection = connection
        End Sub

        Public Sub New(ByVal cmdText As String, ByVal connection As TemplateConnection, ByVal txn As TemplateTransaction)
            MyBase.New()

            m_sCmdText = cmdText
            m_connection = connection
            m_txn = txn
        End Sub

        '****
        '* IMPLEMENT THE REQUIRED PROPERTIES.
        '****
        Public Property CommandText() As String Implements IDbCommand.CommandText
            Get
                Return m_sCmdText
            End Get
            Set(ByVal Value As String)
                m_sCmdText = value
            End Set
        End Property

        Public Property CommandTimeout() As Integer Implements IDbCommand.CommandTimeout
            '*
            ' * The sample does not support a command time-out. As a result,
            ' * for the get, zero is returned because zero indicates an indefinite
            ' * time-out period. For the set, throw an exception.
            '*
            Get
                Return 0
            End Get
            Set(ByVal Value As Integer)
                If value <> 0 Then Throw New NotSupportedException
            End Set
        End Property

        Public Property CommandType() As CommandType Implements IDbCommand.CommandType
            '*
            '* The sample only supports CommandType.Text.
            '*
            Get
                Return CommandType.Text
            End Get
            Set(ByVal Value As CommandType)
                If value <> CommandType.Text Then Throw New NotSupportedException
            End Set
        End Property

        Public Property Connection() As IDbConnection Implements IDbCommand.Connection
            '**
            '* The user should be able to set or change the connection at 
            '* any time.
            '*
            Get
                Return m_connection
            End Get
            Set(ByVal Value As IDbConnection)
                '*
                '* Because the connection is associated with the transaction,
                '* setthe transaction object to return a null reference if the connection 
                '* is reset.
                '*
                If Not m_connection Is value Then Me.Transaction = Nothing

                m_connection = CType(value, TemplateConnection)
            End Set
        End Property

        Protected ReadOnly Property Parameters() As IDataParameterCollection Implements IDbCommand.Parameters
            Get
                Return CType(m_parameters, TemplateParameterCollection)
            End Get
        End Property

        Public Property Transaction() As IDbTransaction Implements IDbCommand.Transaction
            '*
            '* Set the transaction. Consider additional steps to ensure that the transaction
            '* is compatible with the connection, because the two are usually linked.
            '*
            Get
                Return m_txn
            End Get
            Set(ByVal Value As IDbTransaction)
                m_txn = CType(value, TemplateTransaction)
            End Set
        End Property

        Public Property UpdatedRowSource() As UpdateRowSource Implements IDbCommand.UpdatedRowSource
            Get
                Return m_updatedRowSource
            End Get
            Set(ByVal Value As UpdateRowSource)
                m_updatedRowSource = value
            End Set
        End Property

        '****
        '* IMPLEMENT THE REQUIRED METHODS.
        '****
        Public Sub Cancel() Implements IDbCommand.Cancel
            ' The sample does not support canceling a command
            ' once it has been initiated.
            Throw New NotSupportedException
        End Sub

        Public Function CreateParameter() As IDbDataParameter Implements IDbCommand.CreateParameter
            Return CType(New TemplateParameter, IDbDataParameter)
        End Function

        Public Function ExecuteNonQuery() As Integer Implements IDbCommand.ExecuteNonQuery
            '*
            '* ExecuteNonQuery is intended for commands that do
            '* not return results, instead returning only the number
            '* of records affected.
            '*

            ' There must be a valid and open connection.
            If m_connection Is Nothing Or m_connection.State <> ConnectionState.Open Then
                Throw New InvalidOperationException("Connection must valid and open")
            End If

            ' Execute the command.
            Dim resultSet As New SampleDb.SampleDbResultSet()
            'm_connection.SampleDb.Execute(m_sCmdText, resultSet)

            ' Return the number of records affected.
            Return resultSet.recordsAffected
        End Function

        Public Overloads Function ExecuteReader() As IDataReader Implements IDbCommand.ExecuteReader
            '*
            '* ExecuteReader should retrieve results from the data source
            '* and return a DataReader that allows the user to process 
            '* the results.
            '*
            ' There must be a valid and open connection.
            If m_connection Is Nothing Or m_connection.State <> ConnectionState.Open Then
                Throw New InvalidOperationException("Connection must valid and open")
            End If

            ' Execute the command.
            Dim resultSet As SampleDb.SampleDbResultSet = New SampleDb.SampleDbResultSet
            m_connection.SampleDb.Execute(m_sCmdText, resultSet)

            Return New TemplateDataReader(resultSet)
        End Function

        Public Overloads Function ExecuteReader(ByVal behavior As CommandBehavior) As IDataReader Implements IDbCommand.ExecuteReader
            '*
            '* ExecuteReader should retrieve results from the data source
            '* and return a DataReader that allows the user to process 
            '* the results.
            '*

            ' There must be a valid and open connection.
            If m_connection Is Nothing Or m_connection.State <> ConnectionState.Open Then
                Throw New InvalidOperationException("Connection must valid and open")
            End If

            ' Execute the command.
            Dim resultSet As SampleDb.SampleDbResultSet = New SampleDb.SampleDbResultSet
            m_connection.SampleDb.Execute(m_sCmdText, resultSet)

            '*
            '* The only CommandBehavior option supported by this
            '* sample is the automatic closing of the connection
            '* when the user is done with the reader.
            '*
            If behavior = CommandBehavior.CloseConnection Then
                Return New TemplateDataReader(resultSet, m_connection)
            Else
                Return New TemplateDataReader(resultSet)
            End If
        End Function

        Public Function ExecuteScalar() As Object Implements IDbCommand.ExecuteScalar
            '*
            '* ExecuteScalar assumes that the command will return a single
            '* row with a single column, or if more rows/columns are returned
            '* it will return the first column of the first row.
            '*

            ' There must be a valid and open connection.
            If m_connection Is Nothing Or m_connection.State <> ConnectionState.Open Then
                Throw New InvalidOperationException("Connection must valid and open")
            End If

            ' Execute the command.
            Dim resultSet As SampleDb.SampleDbResultSet = New SampleDb.SampleDbResultSet
            m_connection.SampleDb.Execute(m_sCmdText, resultSet)

            ' Return the first column of the first row.
            ' Return a null reference if there is no data.
            If resultSet.data.Length = 0 Then Return Nothing

            Return resultSet.data(0, 0)
        End Function

        Public Sub Prepare() Implements IDbCommand.Prepare
            ' The sample Prepare is a no-op.
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            Me.Dispose(True)
            System.GC.SuppressFinalize(Me)
        End Sub

        Private Sub Dispose(ByVal disposing As Boolean)
            '
            ' Dispose of the object and perform any cleanup.
            '
        End Sub

    End Class
End Namespace

File: src/NBdbm/CS/Classes/Fachada_Net/command.cs
using System.Data;
using System.Collections;
using Microsoft.VisualBasic;
using System.Diagnostics;
using System;



namespace NBdbm
{
	namespace DotNetDataProviderTemplate
	{
		public class TemplateCommand : IDbCommand
		{
			
			
			private TemplateConnection m_connection;
			private TemplateTransaction m_txn;
			private string m_sCmdText;
			private UpdateRowSource m_updatedRowSource = UpdateRowSource.None;
			private TemplateParameterCollection m_parameters = new TemplateParameterCollection();
			
			// Implement the default constructor here.
			public TemplateCommand()
			{
			}
			
			// Implement other constructors here.
			public TemplateCommand(string cmdText)
			{
				
				m_sCmdText = cmdText;
			}
			
			public TemplateCommand(string cmdText, TemplateConnection connection)
			{
				
				m_sCmdText = cmdText;
				m_connection = connection;
			}
			
			public TemplateCommand(string cmdText, TemplateConnection connection, TemplateTransaction txn)
			{
				
				m_sCmdText = cmdText;
				m_connection = connection;
				m_txn = txn;
			}
			
			//****
			//* IMPLEMENT THE REQUIRED PROPERTIES.
			//****
			public string CommandText
			{
				get
				{
					return m_sCmdText;
				}
				set
				{
					m_sCmdText = value;
				}
			}
			
			public int CommandTimeout
			{
				//*
				// * The sample does not support a command time-out. As a result,
				// * for the get, zero is returned because zero indicates an indefinite
				// * time-out period. For the set, throw an exception.
				//*
				get
				{
					return 0;
				}
				set
				{
					if (value != 0)
					{
						throw (new NotSupportedException());
					}
				}
			}
			
			public CommandType CommandType
			{
				//*
				//* The sample only supports CommandType.Text.
				//*
				get
				{
					return CommandType.Text;
				}
				set
				{
					if (value != CommandType.Text)
					{
						throw (new NotSupportedException());
					}
				}
			}
			
			public IDbConnection Connection
			{
				//**
				//* The user should be able to set or change the connection at
				//* any time.
				//*
				get
				{
					return m_connection;
				}
				set
				{
					//*
					//* Because the connection is associated with the transaction,
					//* setthe transaction object to return a null reference if the connection
					//* is reset.
					//*
					if (m_connection != value)
					{
						this.Transaction = null;
					}
					
					m_connection = (TemplateConnection) value;
				}
			}
			
			public IDataParameterCollection Parameters
			{
				get
				{
					return m_parameters;
				}
			}
			
			public IDbTransaction Transaction
			{
				//*
				//* Set the transaction. Consider additional steps to ensure that the transaction
				//* is compatible with the connection, because the two are usually linked.
				//*
				get
				{
					return m_txn;
				}
				set
				{
					m_txn = (TemplateTransaction) value;
				}
			}
			
			public UpdateRowSource UpdatedRowSource
			{
				get
				{
					return m_updatedRowSource;
				}
				set
				{
					m_updatedRowSource = value;
				}
			}
			
			//****
			//* IMPLEMENT THE REQUIRED METHODS.
			//****
			public void Cancel()
			{
				// The sample does not support canceling a command
				// once it has been initiated.
				throw (new NotSupportedException());
			}
			
			public IDbDataParameter CreateParameter()
			{
				return ((IDbDataParameter) (new TemplateParameter()));
			}
			
			public int ExecuteNonQuery()
			{
				//*
				//* ExecuteNonQuery is intended for commands that do
				//* not return results, instead returning only the number
				//* of records affected.
				//*
				
				// There must be a valid and open connection.
				if (m_connection == null || m_connection.State != ConnectionState.Open)
				{
					throw (new InvalidOperationException("Connection must valid and open"));
				}
				
				// Execute the command.
				SampleDb.SampleDbResultSet resultSet = new SampleDb.SampleDbResultSet();
				//m_connection.SampleDb.Execute(m_sCmdText, resultSet)
				
				// Return the number of records affected.
				return resultSet.recordsAffected;
			}
			
			public IDataReader ExecuteReader()
			{
				//*
				//* ExecuteReader should retrieve results from the data source
				//* and return a DataReader that allows the user to process
				//* the results.
				//*
				// There must be a valid and open connection.
				if (m_connection == null || m_connection.State != ConnectionState.Open)
				{
					throw (new InvalidOperationException("Connection must valid and open"));
				}
				
				// Execute the command.
				SampleDb.SampleDbResultSet resultSet = new SampleDb.SampleDbResultSet();
				m_connection.SampleDb.Execute(m_sCmdText, ref resultSet);
				
				return new TemplateDataReader(resultSet);
			}
			
			public IDataReader ExecuteReader(CommandBehavior behavior)
			{
				//*
				//* ExecuteReader should retrieve results from the data source
				//* and return a DataReader that allows the user to process
				//* the results.
				//*
				
				// There must be a valid and open connection.
				if (m_connection == null || m_connection.State != ConnectionState.Open)
				{
					throw (new InvalidOperationException("Connection must valid and open"));
				}
				
				// Execute the command.
				SampleDb.SampleDbResultSet resultSet = new SampleDb.SampleDbResultSet();
				m_connection.SampleDb.Execute(m_sCmdText, ref resultSet);
				
				//*
				//* The only CommandBehavior option supported by this
				//* sample is the automatic closing of the connection
				//* when the user is done with the reader.
				//*
				if (behavior == CommandBehavior.CloseConnection)
				{
					return new TemplateDataReader(resultSet, m_connection);
				}
				else
				{
					return new TemplateDataReader(resultSet);
				}
			}
			
			public object ExecuteScalar()
			{
				//*
				//* ExecuteScalar assumes that the command will return a single
				//* row with a single column, or if more rows/columns are returned
				//* it will return the first column of the first row.
				//*
				
				// There must be a valid and open connection.
				if (m_connection == null || m_connection.State != ConnectionState.Open)
				{
					throw (new InvalidOperationException("Connection must valid and open"));
				}
				
				// Execute the command.
				SampleDb.SampleDbResultSet resultSet = new SampleDb.SampleDbResultSet();
				m_connection.SampleDb.Execute(m_sCmdText, ref resultSet);
				
				// Return the first column of the first row.
				// Return a null reference if there is no data.
				if (resultSet.data.Length == 0)
				{
					return null;
				}
				
				return resultSet.data[0, 0];
			}
			
			public void Prepare()
			{
				// The sample Prepare is a no-op.
			}
			
			public void Dispose()
			{
				this.Dispose(true);
				System.GC.SuppressFinalize(this);
			}
			
			private void Dispose(bool disposing)
			{
				//
				// Dispose of the object and perform any cleanup.
				//
			}
			
		}
	}
	
	
}
